File: apps/sync/test/chalk_sync/whiteboard_v1/reducer_test.exs
defmodule ChalkSync.WhiteboardV1.ReducerTest do
  use ExUnit.Case, async: true

  alias ChalkSync.WhiteboardV1.Reducer

  @fixture_path Path.expand(
                  "../../../../../packages/whiteboard/src/collab/fixtures/excalidraw-0.18.1-reducer-golden.json",
                  __DIR__
                )

  test "matches the Excalidraw 0.18.1 reducer golden" do
    fixture = @fixture_path |> File.read!() |> JSON.decode!()

    assert Reducer.merge(fixture["current"], fixture["incoming"]) == fixture["expected"]
  end

  test "a full sync does not delete absent elements" do
    fixture = @fixture_path |> File.read!() |> JSON.decode!()
    assert Reducer.merge(fixture["current"], []) == fixture["current"]
  end
end

File: packages/whiteboard/src/collab/reducer.test.ts
import { describe, expect, it } from "vitest";

import fixture from "./fixtures/excalidraw-0.18.1-reducer-golden.json";
import { mergeWhiteboardElements } from "./reducer";
import type { WhiteboardWireElement } from "./wire";

describe("whiteboard-v1 deterministic reducer", () => {
  it("matches the Excalidraw 0.18.1 golden merge", () => {
    expect(mergeWhiteboardElements(fixture.current as WhiteboardWireElement[], fixture.incoming as WhiteboardWireElement[])).toEqual(fixture.expected);
  });

  it("does not infer deletions from absent elements in a full sync", () => {
    expect(mergeWhiteboardElements(fixture.current as WhiteboardWireElement[], [])).toEqual(fixture.current);
  });
});
